compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,231 @@
+package dotty.tools
+package dotc
+package transform
+
+import TreeTransforms.{ MiniPhaseTransform, TransformerInfo }
+import core._
+import Contexts.Context, Types._, Decorators._, Symbols._, DenotTransformers._
+import Denotations._, SymDenotations._, Scopes._, StdNames._, NameOps._, Names._
+import ast.tpd
+
+class SpecializeFunctions extends MiniPhaseTransform with DenotTransformer {
+  import ast.tpd._
+
+  val phaseName = "specializeFunction1"
+
+  // Setup ---------------------------------------------------------------------
+  private[this] val functionName = "JFunction".toTermName
+  private[this] val functionPkg  = "scala.compat.java8.".toTermName
+  private[this] var argTypes: Set[Symbol] = _
+  private[this] var retTypes: Set[Symbol] = _
+
+  override def prepareForUnit(tree: Tree)(implicit ctx: Context) = {
+    retTypes = Set(defn.UnitClass,
+                   defn.BooleanClass,
+                   defn.IntClass,
+                   defn.FloatClass,
+                   defn.LongClass,
+                   defn.DoubleClass,
+                   /* only for Function0: */
+                   defn.ByteClass,
+                   defn.ShortClass,
+                   defn.CharClass)
+
+    argTypes = Set(defn.IntClass,
+                   defn.LongClass,
+                   defn.DoubleClass,
+                   /* only for Function1: */
+                   defn.FloatClass)
+    this
+  }
+
+  // Transformations -----------------------------------------------------------
+
+  /** Transforms all classes extending `Function1[-T1, +R]` so that
+    * they instead extend the specialized version `JFunction$mp...`
+    */
+  def transform(ref: SingleDenotation)(implicit ctx: Context) = ref match {
+    case cref @ ShouldTransformDenot(targets) => {
+      val specializedSymbols: Map[Symbol, (Symbol, Symbol)] = (for (SpecializationTarget(target, args, ret, original) <- targets) yield {
+        val arity = args.length
+        val specializedParent = ctx.getClassIfDefined {
+          functionPkg ++ specializedName(functionName ++ arity, args, ret)
+        }
+
+        val specializedApply: Symbol = {
+          val specializedMethodName = specializedName(nme.apply, args, ret)
+          ctx.newSymbol(
+            cref.symbol,
+            specializedMethodName,
+            Flags.Override | Flags.Method,
+            specializedParent.info.decls.lookup(specializedMethodName).info
+          )
+        }
+
+        original -> (specializedParent, specializedApply)
+      }).toMap
+
+      def specializeApplys(scope: Scope): Scope = {
+        val alteredScope = scope.cloneScope
+        specializedSymbols.values.foreach { case (_, apply) =>
+          alteredScope.enter(apply)
+        }
+        alteredScope
+      }
+
+      def replace(in: List[TypeRef]): List[TypeRef] =
+        in.map { tref =>
+          val sym = tref.symbol
+          specializedSymbols.get(sym).map { case (specializedParent, _) =>
```

Comment:
I this this is a very inefficient implementation.

Instead of building a `Seq[(Symbol, (Symbol, Symbol)]` and then converting it and operating it,
you can use the fact that both `specializedSymbols` and `parents` maintain the same order and traverse them together.

This will allow for an implementation that doesn't allocate any tuples, doesn't convert collections and traverses collections only once.